### ISSUE TITLE
RFC: Use `offsetKey`s to specify labware offsets

### ISSUE THREAD

**Issue by SyntaxColoring** [Opentrons/opentrons#9354]
# Background and motivation

As of v5.0.0-beta.7, the HTTP API for setting up a run with labware offsets works as described in #9351.

We (the CPX and App+UI squads) have at least two problems with this:

1. When the client wants to `POST` an offset, it has to blindly copy a handful of fields from the analysis that the server returned, and input them back to the server. This is basically just busy work.
2. More importantly, *one* of those fields, `moduleModel`, is nuanced and pernicious.

   ~~The client can't simply copy *this* field from the analysis, because of how the server handles module compatibility. (See #9351 for details.) Instead, the client must take pains to predict what the `moduleModel` will be during the actual run (which will not necessarily match analysis), and send *that* to the server.~~

   [Edit: This is no longer true with the contract changes in #9364.]

If the client or server gets any of this wrong, the failure mode is pretty bad: the protocol could silently run with the wrong labware offsets.

# Proposal for API changes

The `result` of a `loadLabware` command will gain a new field, `offsetKey`. It will be an opaque string chosen by the server. It will be present in both analysis and actual execution.

`POST /runs/{id}/labware_offsets` will change to look like this:

```javascript
{
	// Remains the same:
	"vector": { "x": 1, "y": 2, "z": 3 }
	
	// New:
	"offsetKey": "abc123def456",
	
	// Deleted:
	// "definitionUri": "..."
	// "location": {"slotName": "1", "moduleModel": "temperatureModuleV1"}
}
```

(We'll also make matching changes to the other places where this labware offset model shows up. At the time of writing, that's `GET /runs` and `GET /runs/{id}`.)

Our flow of providing a labware offset will then become, at a high level:

1. The client uploads a protocol and creates a run from it.
2. The client looks at the protocol's analysis and identifies what labware, modules, and pipettes the protocol is anticipated to use.
3. The client uses live Protocol Engine commands to prepare the deck and measure labware offsets however it pleases.
4. The client `POST`s the measured vectors, **with the `offsetKey`s from the protocol's analysis.**
5. The client starts the protocol run.
6. When the protocol loads a labware...
   1. The server checks to see if this load matches one that it anticipated in analysis.\*
   2. **If yes, then it looks to see if a labware offset was `POST`ed with that anticipated labware's `offsetKey`.**
   3. If yes, it applies that offset vector.

\*What counts as a "match" is left up to the server. But the implementation will probably compare definition URIs and locations, just like we currently do.

It only makes sense for a client to `POST` a labware offset with an `offsetKey` that was in the protocol's analysis. Possibly, the server should return an error if the client accidentally uses something else.

# Upsides

* This would greatly simplify the bidirectional interplay between the client and server, and make things easier for us to understand and get right. This means fewer bugs for a sensitive part of the product.
* It would let us rely more on tests that are simpler.

   Most of our tests will treat the client and server in isolation. Those tests can check that the client or server uses the API "correctly," but they can't check that the test itself has a good definition of "correct." With the current API, there's a big gap "in between" the client and server where misunderstandings can subtly break things without failing tests. The proposed API leaves less room in that gap.
* It would give us flexibility to evolve what it means for a labware in a run to "match" a labware that was predicted in analysis. This will be important if we flesh out our support for protocols that change their deck layout midway through.

   With the current API, every evolution would affet the API and necessitate client-server cooperation. With the proposed API, the changes can be made as implementation details on the server. 

# Downsides and why I think they're okay

## Can't take advantage of side-channel labware knowledge

This technically takes some power away from the API.

Suppose you're a 3rd-party integrator, and you have a protocol that you *know* will load a certain labware in a certain location. But for whatever reason, that load doesn't show up in the protocol's analysis.

The current API lets you `POST` a functioning labware offset despite this. The proposed API would not.

In my opinion, this is fine, and maybe even a good thing. We don't know that the current fields of `definitionUri`, `deckSlotName`, and `moduleModel` are what a 3rd-party integrator would want the match to be based on. Conceivably, they might want to do any of these, instead:

* Apply an offset to the *n*th labware that gets loaded, regardless of what kind of labware it is or where it's loaded.
* Apply an offset to whatever labware gets loaded in slot *n*.
* Apply an offset to whatever labware gets loaded with the custom label `"foo"`.

In other words, `definitionUri`+`deckSlotName`+`moduleModel` were kind of arbitrary to begin with. So until we've investigated what 3rd-party integrators would want, I don't think we should feel like we have to preserve the current API for their sake.

**Comment by SyntaxColoring** [comment]
Here's one possible way to implement this on the server side:

1. In analysis, build `offsetKey`s as concatenations of the `definitionUri`, `slotName`, and the protocol's *requested* `moduleModel`. For example, `"opentrons/my_labware/1-1-temperatureModuleV1"`.
2. In the actual run, build the same concatenation, taking care to use the *requested* `moduleModel` instead of the *actually loaded* `moduleModel`.
3. Look for any `POST`ed offsets with that `offsetKey`.

**Comment by SyntaxColoring** [comment]
**We're rejecting this for now,** having discussed this with @mcous and @b-cooper individually.

* Things are working in production, so far. We haven't seen any bugs relating to the client `POST`ing a labware offset with the wrong `definitionUri` or `location`. 🤞 
* The contract changes in #9364 make `location` way less of a trap, taking away maybe 60% of the motivation for these changes.
* These are new concepts that we'd all have to learn.
* We're not confident that the paradigm behind these new concepts will hold water.
